retro/reco.py

```diff
@@ -509,20 +694,133 @@ def loglike(cube, ndim=None, nparams=None): # pylint: disable=unused-argument
+        '''
+        # Setup LLHP dtype
+        dim_names = list(self.hypo_handler.all_param_names)
+
+        # add derived quantities
+        derived_dim_names = ['energy', 'azimuth', 'zenith']
+        if 'cascade_d_zenith' in dim_names and 'cascade_d_azimuth' in dim_names:
+            derived_dim_names += ['cascade_zenith', 'cascade_azimuth']
+
+        all_dim_names = dim_names + derived_dim_names
+
+        llhp_t = np.dtype([(field, np.float32) for field in ['llh'] + all_dim_names])
+
+        # dump
+        llhp = np.zeros(shape=len(return_param_values), dtype=llhp_t)
+        llhp['llh'] = return_log_likelihoods
+        llhp[dim_names] = return_param_values
+
+        
+        # create derived dimensions
+        if 'energy' in derived_dim_names:
+            if 'track_energy' in dim_names:
+                llhp['energy'] += llhp['track_energy']
+            if 'cascade_energy' in dim_names:
+                llhp['energy'] += llhp['cascade_energy']
+
+        if 'cascade_d_zenith' in dim_names and 'cascade_d_azimuth' in dim_names:
+            # create cascade angles from delta angles
+            rotate_point(p_theta = llhp['cascade_d_zenith'], p_phi = llhp['cascade_d_azimuth'],
+                         rot_theta = llhp['track_zenith'], rot_phi = llhp['track_azimuth'],
+                         q_theta = llhp['cascade_zenith'], q_phi = llhp['cascade_azimuth'])
+
+        if 'track_zenith' in all_dim_names and 'track_azimuth' in all_dim_names:
+            if 'cascade_zenith' in all_dim_names and 'cascade_azimuth' in all_dim_names:
+                # this resulting radius we won't need, but need to supply an array to the function
+                r_out = np.empty(shape=llhp.shape, dtype=np.float32)
+                # combine angles:
+                add_vectors(r1 = llhp['track_energy'], theta1 = llhp['track_zenith'], phi1 = llhp['track_azimuth'],
+                            r2 = llhp['cascade_energy'], theta2 = llhp['cascade_zenith'], phi2 = llhp['cascade_azimuth'],
+                            r3 = r_out, theta3 = llhp['zenith'], phi3 = llhp['azimuth'])
+            else:
+                # in this case there is no cascade angles
+                llhp['zenith'] = llhp['track_zenith']
+                llhp['azimuth'] = llhp['track_azimuth']
+
+        elif 'cascade_zenith' in all_dim_names and 'cascade_azimuth' in all_dim_names:
+            # in this case there are no track angles
+            llhp['zenith'] = llhp['cascade_zenith']
+            llhp['azimuth'] = llhp['cascade_azimuth']
+
+        # save full info
+        if fname is not None:
+            llhp_outf = self.out_prefix + fname + '.npy'
+            print('Saving llhp to "{}"...'.format(llhp_outf))
+            np.save(llhp_outf, llhp)
 
-    def run_test(self, prior, loglike):
+        return llhp
+
+    def make_meta_dict(self, settings, llhp=None, time=-1, fname=None):
+        '''
+        create meta information dictionary
+        '''
```

Comment:
document args; and the above does not indicate that it also saves the dict if you provide the `fname` arg, so e.g.
```python
"""Create meta information dictionary and optionally save to disk.

Parameters
----------
blah : ...
```
also note that the description should be on the same line with the opening quotes of the docstring (numpy convention docstrings)